utils/indicators.py

```diff
@@ -0,0 +1,304 @@
+
+        logging.info(f"Generated index: {index_config['name']}")
+        return dataset
+
+    def zonal_mean_index(self, index_key, year):
+        index_config = self.indices[index_key]
+        dataset = self.generate_index(index_config, year)
+        # zm = self._zonal_mean(single, index_config.get('bandname') or 'constant')
+        out = dataset.reduceRegion(
+            **{
+                "reducer": ee.Reducer.mean(),
+                "geometry": self.roi,
+                "scale": 200,  # map scale
+            }
+        ).getInfo()
+        if index_config.get("bandname"):
+            return out[index_config.get("bandname")]
+        return out
+
+    def generate_composite_index_df(self, year, project_geometry, indices=[]):
+        
+        data = {
+            "metric": indices,
+            "year": year,
+            "centroid": "",
+            "project_name": "",
+            "value": list(map(self.zonal_mean_index, indices, repeat(year))),
+            "area": self.roi.area().getInfo(),  # m^2 to-do: calculate with duckdb
+            "geojson": "",
+            # to-do: coefficient
+        }
+
+        logging.info("data", data)
+        df = pd.DataFrame(data)
+        return df
+
+    @staticmethod
+    def _authenticate_ee(ee_service_account):
+        """
+        Huggingface Spaces does not support secret files, therefore authenticate with an environment variable containing the JSON.
+        """
+        logging.info("Authenticating to Google Earth Engine...")
+        credentials = ee.ServiceAccountCredentials(
+            ee_service_account, key_data=os.environ["ee_service_account"]
+        )
+        ee.Initialize(credentials)
+        logging.info("Authenticated to Google Earth Engine.")
+
+    def _calculate_yearly_index(self, years, project_name):
+        dfs = []
+        logging.info(years)
+        project_geometry = dq.get_project_geometry(project_name)
+        project_centroid = dq.get_project_centroid(project_name)
+        # to-do: refactor to involve less transformations
+        _polygon = json.dumps(json.loads(project_geometry[0][0])['features'][0]['geometry'])
+        # to-do: don't use self.roi and instead pass patameter strategically
+        self.roi = ee.Geometry.Polygon(json.loads(_polygon)['coordinates'])
+
+        # to-do: pararelize?
+        for year in years:
+            logging.info(year)
+            self.project_name = project_name
+            df = self.generate_composite_index_df(year, project_geometry, list(self.indices.keys()))
+            dfs.append(df)
+
+        # Concatenate all dataframes
+        df_concat = pd.concat(dfs)
+        df_concat['centroid'] = project_centroid
+        df_concat['project_name'] = project_name
+        df_concat['geojson'] = project_geometry
+        breakpoint()
+        return df_concat
+
+    # h/t: https://community.plotly.com/t/dynamic-zoom-for-mapbox/32658/12
+    def _latlon_to_config(self, longitudes=None, latitudes=None):
```

Comment:
If this (renamed) function exists in service of centering the map on the polygon, it's possible it can be replaced by `ee`'s native functionality to center map on polygon. @buuck

Reply:
The primary utility of that function was to set the mapbox zoom level automatically. Is that something we can get from `ee`?